agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -0,0 +1,233 @@
+            webSocket.waitForMessageOfType<RequestReceivedMessage>(MESSAGE_TYPES.REQUEST_RECEIVED).also {
+                counterParty = IndyParty(it.did, it.endpoint)
+                sendJson(RequestSendResponseMessage(it.did))
+            }
+        } else {
+            throw AgentConnectionException("Agent is disconnected")
+        }
+    }
+
+    fun sendRequest(key: String) = webSocket.sendJson(SendRequestMessage(key))
+
+    override fun getCounterParty(): IndyParty? {
+        return counterParty
+    }
+
+    override fun sendCredentialOffer(offer: CredentialOffer) = webSocket.sendTypedMessage(offer, counterParty!!)
+
+    override fun receiveCredentialOffer(): CredentialOffer = webSocket.waitForTypedMessage()
+
+    override fun sendCredentialRequest(request: CredentialRequestInfo) = webSocket.sendTypedMessage(request, counterParty!!)
+
+    override fun receiveCredentialRequest(): CredentialRequestInfo = webSocket.waitForTypedMessage()
+
+    override fun sendCredential(credential: CredentialInfo) = webSocket.sendTypedMessage(credential, counterParty!!)
+
+    override fun receiveCredential(): CredentialInfo = webSocket.waitForTypedMessage()
+    override fun sendProofRequest(request: ProofRequest) = webSocket.sendTypedMessage(request, counterParty!!)
+
+    override fun receiveProofRequest(): ProofRequest = webSocket.waitForTypedMessage()
+
+    override fun sendProof(proof: ProofInfo) = webSocket.sendTypedMessage(proof, counterParty!!)
+
+    override fun receiveProof(): ProofInfo = webSocket.waitForTypedMessage()
+}
+
+class AgentWebSocketClient(serverUri: URI) : WebSocketClient(serverUri) {
+    private val log = KotlinLogging.logger {}
+
+    override fun onOpen(handshakedata: ServerHandshake?) {
+        log.info { "AgentConnection opened: $handshakedata" }
+    }
+
+    override fun onClose(code: Int, reason: String?, remote: Boolean) {
+        log.info { "AgentConnection closed: $code,$reason,$remote" }
+    }
+
+    val receivedMessages = mutableListOf<String>()
+
+    override fun onMessage(message: String?) {
+        log.info { "Message: $message" }
+        if (message != null)
+            synchronized(receivedMessages) { receivedMessages.add(message) }
+    }
+
+    override fun onError(ex: Exception?) {
+        log.warn(ex) { "AgentConnection error" }
+    }
+
+    fun sendJson(obj: Any) = send(SerializationUtils.anyToJSON(obj))
+
+    fun popMessageOfType(type: String): String? {
+        synchronized(receivedMessages) {
+            val result = receivedMessages.find { SerializationUtils.jSONToAny<Map<String, Any>>(it)["@type"].toString().contentEquals(type) }
+            if (result != null)
+                receivedMessages.remove(result)
+
+            return result
+        }
+    }
+
+    fun waitForMessageOfType(type: String): String {
+        var messageOfType: String? = null
+        while (messageOfType == null) {
+            sleep(500)
```

Comment:
Need to re-work blocking code.
Let's put it in backlog and mention in driver limitation *.md

Reply:
https://github.com/hyperledger-labs/cordentity/projects/1#card-18735696